hibernate-core/src/main/java/org/hibernate/query/internal/AbstractProducedQuery.java

```diff
@@ -749,57 +749,86 @@ public boolean isBound(Parameter<?> parameter) {
+		final String boundCheckingLogMessage = "Checking whether parameter reference " + parameter + " is bound : %s";
 
-		final QueryParameterBinding<T> binding = getQueryParameterBindings().getBinding( (QueryParameter<T>) parameter );
-		LOGGER.debugf( "Checking whether parameter reference [%s] is bound : %s", parameter, binding.isBound() );
-		if ( !binding.isBound() ) {
-			throw new IllegalStateException( "Parameter value not yet bound : " + parameter.toString() );
-		}
-		return binding.getBindValue();
+		return (T) getParameterValue(
+				(QueryParameter) parameter,
+				notBoundErrorMessage,
+				couldNotResolveErrorMessage,
+				boundCheckingLogMessage
+		);
 	}
 
 	@Override
 	public Object getParameterValue(String name) {
 		getProducer().checkOpen( false );
 
-		final QueryParameterBinding binding;
-		try {
-			binding = getQueryParameterBindings().getBinding( name );
-		}
-		catch (QueryParameterException e) {
-			throw new IllegalArgumentException( "Could not resolve parameter by name - " + name, e );
-		}
+		final String notBoundErrorMessage = "Parameter value not yet bound : " + name;
+		final String couldNotResolveErrorMessage = "Could not resolve parameter by name - " + name;
+		final String boundCheckingLogMessage = "Checking whether named parameter [ " + name + "] is bound : %s";
 
-		LOGGER.debugf( "Checking whether named parameter [%s] is bound : %s", name, binding.isBound() );
-		if ( !binding.isBound() ) {
-			throw new IllegalStateException( "Parameter value not yet bound : " + name );
-		}
-		return binding.getBindValue();
+		final QueryParameter<Object> queryParameter = getParameterMetadata().getQueryParameter( name );
+		return getParameterValue(
+				queryParameter,
+				notBoundErrorMessage,
+				couldNotResolveErrorMessage,
+				boundCheckingLogMessage
+		);
 	}
 
 	@Override
 	public Object getParameterValue(int position) {
 		getProducer().checkOpen( false );
 
-		final QueryParameterBinding binding;
+		final String notBoundErrorMessage = "Parameter value not yet bound : " + position;
+		final String couldNotResolveErrorMessage = "Could not resolve parameter by position - " + position;
+		final String boundCheckingLogMessage = "Checking whether positional parameter [ " + position + "] is bound : %s";
+
+		final QueryParameter<Object> queryParameter = getParameterMetadata().getQueryParameter( position );
+		return getParameterValue(
+				queryParameter,
+				notBoundErrorMessage,
+				couldNotResolveErrorMessage,
+				boundCheckingLogMessage
+		);
+	}
+
+	private Object getParameterValue(
+			QueryParameter queryParameter,
+			String notBoundErrorMessage,
+			String couldNotResolveErrorMessage,
+			String boundCheckingLogMessage) {
 		try {
-			binding = getQueryParameterBindings().getBinding( position );
+			final QueryParameterBindings parameterBindings = getQueryParameterBindings();
+
+			if ( queryParameter == null ) {
+				throw new IllegalArgumentException( couldNotResolveErrorMessage );
```

Comment:
Seems the exception is not related to the method parameter. `IllegalStatementException` is more appropriate?

Reply:
if this happens then we were not able to resolve the parameter, that is the argument of the method, plus to to be consistent with what we had before an `IllegalArgumentException` seems correct to me.